compiler/hash-typecheck/src/params.rs

```diff
@@ -0,0 +1,381 @@
+    ArgNameNotFoundInParams { arg: SomeArgId, params: ParamsId },
+    RequiredParamNotFoundInArgs { param: ParamId, args: SomeArgsId },
+    SpreadBeforePositionalArg { next_positional: SomeArgId },
+}
+
+impl<T: AccessToTypechecking> ParamOps<'_, T> {
+    /// Validate the given parameters, returning an error if they are invalid.
+    ///
+    /// Conditions for valid parameters are:
+    /// 1. No duplicate parameter names
+    /// 2. All parameters with defaults are at the end
+    pub fn validate_params(&self, params_id: ParamsId) -> TcResult<()> {
+        let mut error_state = self.new_error_state();
+
+        let mut seen = HashSet::new();
+        let mut found_default = None;
+
+        for param in params_id.iter() {
+            // Check for duplicate parameters
+            if let Some(param_name) = self.get_param_name_ident(param) {
+                if seen.contains(&param_name) {
+                    error_state
+                        .add_error(ParamError::DuplicateParam { first: param, second: param });
+                }
+                seen.insert(param_name);
+            }
+
+            // Ensure that parameters with defaults follow parameters without
+            // defaults
+            if let Some(default_param) = found_default {
+                if self.get_param_default(param).is_none() {
+                    // Required parameter after named parameter,
+                    // error
+                    error_state.add_error(ParamError::RequiredParamAfterDefaultParam {
+                        first_default: default_param,
+                        next_required: param,
+                    });
+                }
+            } else if self.get_param_default(param).is_some() {
+                // Found the first default parameter
+                found_default = Some(param);
+            }
+        }
+
+        self.return_or_register_errors(|| Ok(()), error_state)
+    }
+
+    /// Validate the given arguments against the given parameters, returning an
+    /// error if they are invalid.
+    ///
+    /// Conditions for valid arguments are:
+    /// 1. No duplicate argument names
+    /// 2. All named arguments follow positional arguments
+    /// 3. No more arguments than parameters
+    ///
+    /// The specific unification of the argument and parameter types is not
+    /// checked at this stage. The function
+    /// `validate_and_reorder_args_against_params` performs additional
+    /// validation of the argument names, and reorders the arguments to match
+    /// the parameters.
+    pub fn validate_args_against_params(
+        &self,
+        args_id: SomeArgsId,
+        params_id: ParamsId,
+    ) -> TcResult<()> {
+        let mut error_state = self.new_error_state();
+
+        // Check for too many arguments
+        if args_id.len() > params_id.len() {
+            error_state.add_error(ParamError::TooManyArgs { expected: params_id, got: args_id });
+        }
```

Comment:
seems that this function doesn't deal with defaults yet?

Reply:
It does.